### ISSUE TITLE
Intermittent hang when using UnsafeCallback

### ISSUE THREAD

**Issue by ericsink** [denoland/deno#18898]
My project is using Deno FFI with UnsafeCallback.  I have a test case that hangs around 60-70% of the time.

The problem is reproducible with 1.32.5 and with changeset 10ae5ee26557107b22524b1a84ebb56ed7d23fb4 which I just pulled and built.  My machine is a Mac laptop with the M1/arm64 chip.  I have not tried this (yet?) on other platforms.

The hang seems to be happening in this function:

https://github.com/denoland/deno/blob/10ae5ee26557107b22524b1a84ebb56ed7d23fb4/ext/ffi/callback.rs#L119

If I put debugging print statements in that function and build my own Deno I can gather a bit of information about what's happening.

This `if` statement is key:

https://github.com/denoland/deno/blob/10ae5ee26557107b22524b1a84ebb56ed7d23fb4/ext/ffi/callback.rs#L126

when that `if` statement is true, everything works fine.

When it is false, execution falls down to this line:

https://github.com/denoland/deno/blob/10ae5ee26557107b22524b1a84ebb56ed7d23fb4/ext/ffi/callback.rs#L130

and as far as I can tell, it never succeeds when that path happens.  Comments in that block of code speak about blocking.

The specific hanging line is:

https://github.com/denoland/deno/blob/10ae5ee26557107b22524b1a84ebb56ed7d23fb4/ext/ffi/callback.rs#L145

Every time the hang happens, it reaches that line but never gets past it.  In those cases, additional debugging print statements in this function never seem be reached:

https://github.com/denoland/deno/blob/10ae5ee26557107b22524b1a84ebb56ed7d23fb4/ext/ffi/callback.rs#L150

I am quite unfamiliar with Deno's code.  I cannot be certain this is a bug in Deno, but I'm logging this issue because it seems like it *might* be.  Or perhaps I am doing something wrong.  Not sure.

**Comment by aapoalas** [comment]
Hi, thanks for the report!

What's happening here is that the first if is checking if your callback is happening synchronously on the JavaScript event loop thread or if it's coming from a foreign thread.

The else path is the call coming from another thread. In this case what you need is a thread safe callback. Thread safe callbacks in Deno are different in normal FFI callbacks in only one regard: Deno expects that thread safe callbacks might be called at any time, even when no JavaScript work is happening, and prepares a way for these thread safe callbacks to wake up the event loop on when called.

A thread safe callback is either created with a reference count of 1 using `Deno.UnsafeCallback.threadSafe()` or from an existing callback by calling its `ref()` method, which increments its reference count by one. Once a callback's reference count has become non-zero it is a thread safe callback and will always remain one. While the ref count of a callback is non-zero it will keep Deno's event loop from exiting. To put it another way, while the ref count is non-zero Deno will keep waiting for new calls to come in from the FFI callback.

So: It seems to me that the reason your test hangs is that the callback has never had a reference count that is non-zero, and thus Deno doesn't expect any incoming calls from it. Something else seems to be keeping your test still running, but nothing is actually making the event loop spin. When the event loop spins, incoming calls would be checked anyway, be they thread safe or not.

Adding `ref()` calls where appropriate (and `unref()` after you're done with the callback) should make the callback capable of waking up the event loop when it gets called.

**Comment by ericsink** [comment]
Thanks for the reply.

However...

In my initial report, I forgot to mention that one of the things I tried was adding a `ref()` call, and it seemed to make no difference.

Just to be sure, I just tried it again.  Same basic result.  With the `ref()` call immediately after I create the callback, I still get the intermittent hang behavior.  In the cases, where it does succeed, I now get an error later, because I never called `unref()`, but that should be expected.

I had not previously tried using `Deno.UnsafeCallback.threadSafe()`, but I did that just now, and (unsurprisingly) it produces the same behavior.

More information, in case it is diagnostically helpful:  

I'm using `deno test`.  I have not tried this in the (probably simpler) context of `deno run`.

When I specify `--parallel`, the hang seems to happen less often (which is the exact opposite of my intuition), but it does still sometimes happen.

**Comment by ericsink** [comment]
Hmmm.

I don't think there are any threads in play, but let's say I'm only about 95% sure of that.  My code involves no threading or async stuff at all.  But what I'm doing here is generating bindings for another library.  The stuff I'm calling in that other library seems very unlikely to be creating a thread, but I haven't actually looked under the hood to be sure.

Still, your question got me thinking.

My little test "suite" has 11 tests across 6 files.  The file containing the hanging test is called `list_test.ts`.

In general, I've been running them all, using:

```
deno test --check --allow-ffi --unstable
```

But if I run the tests in just the file containing the hanging test:

```
deno test --check --allow-ffi --unstable list_test.ts
```

It succeeds.  I ran it 100 times in a row with a script, no hangs, no failures.

Hmmm.  Let's try adding a second file, `string_test.ts`, which does not use any callbacks:

```
deno test --check --allow-ffi --unstable list_test.ts string_test.ts
```

Same result here.  100 consecutive runs with no hangs.

However, if I try the same thing with two files that both contain tests using callbacks:

```
deno test --check --allow-ffi --unstable list_test.ts func_test.ts
```

The hangs are back.

Now, just for fun, let's move a test (which uses a callback) from `func_test.ts` into `list_test.ts`, and run just that file.  Now I can once again do 100 runs with no hangs.

Bottom line:  From what I can tell, the problem only happens when two or more test files are involved and each of them has something where a callback is involved.

**Comment by aapoalas** [comment]
As said in my first reply, the first if you were looking at checks if the callback is being called from a foreign thread: Since the code is going to the else branch, that means there's 100% threading going on.

**Comment by ericsink** [comment]
> there's 100% threading going on

I believe you.  I just don't know where the threading is coming from.  Based on the information I gathered and described in my previous comment, I have to wonder:  Is it possible that it's coming from `deno test`?

The alternative is that something I'm doing is introducing a thread that I don't know about.

I'll try to create a simpler repro, one where I can be 100% sure of what is going on instead of 95%.

**Comment by aapoalas** [comment]
What is the library you're binding to? Can you give that information, and possibly show your code as well?

**Comment by ericsink** [comment]
FWIW, following the conjecture that `Deno test` is introducing threading, I've tried fiddling with `--parallel` and the `DENO_JOBS` environment variable, but no luck.

**Comment by ericsink** [comment]
I've avoided saying too much about what I'm doing simply because I thought the detail would distract.

The "library" is, well, a piece of the .NET 8 class libraries, compiled to native code using "Native AOT", with a generated layer of C-compatible wrapper functions, plus a generated wrapper that exposes things to TypeScript.  There are a lot of pieces in play.

My code is too large to post, so I've been hoping the diagnostic info I've posted might trigger a light bulb on your side.

I mean, do you think it might be relevant that the hang only happens with `deno test` when multiple files are involved?  I don't think I've seen you comment on that particular tidbit.

Anyway, my next step would be to try and construct a simple repro.  If I can reproduce the problem with most of the complexity of my project stripped out, then I've got code I can give you that would be helpful.  And if I can't see the problem with simpler code, then the suspicion of guilt around MY code will grow.

**Comment by ericsink** [comment]
Minimal repro attached.

[issue18898.tar.gz](https://github.com/denoland/deno/files/11357777/issue18898.tar.gz)

Build the `foo.c` file as a shared library and fix its path for your platform in `ffi.ts`.

There are a few tests in `a_test.ts`

With just one copy of that file, they pass every time.

To see the hang, add more copies of the file (such as `b_test.ts`, etc) and run `deno test` again.

**Comment by aapoalas** [comment]
> I mean, do you think it might be relevant that the hang only happens with deno test when multiple files are involved? I don't think I've seen you comment on that particular tidbit.

I would rather it didn't because that will make this quite complicated indeed :) But it's definitely possible. I just don't know enough about `deno test` to say for certain.

**Comment by aapoalas** [comment]
I am not available to debug this for a few days but some interesting things to check are:
1. The `*s.borrow()` value: Is it a null pointer or not? A null pointer would indicate that the call is happening on a thread where no JavaScript has run, or at least no FFI callbacks have been created.
2. If a JS `console.log` is added after each FFI API call used in the hanging tests, do those get called and log? If not, it means that the FFI API triggers the callback and waits for it to finish, indicating a deadlock.

**Comment by aapoalas** [comment]
I was unable to repeat this with newest Deno version, but downgrading to 1.32.5 indeed did immediately show the issue.
